presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddIntermediateAggregations.java

```diff
@@ -224,8 +216,31 @@ private PlanNode addGatheringIntermediate(AggregationNode aggregation, PlanNodeI
                 return ImmutableMap.of();
             }
             VariableReferenceExpression input = getOnlyElement(extractAggregationUniqueVariables(entry.getValue(), types));
-            builder.put(input, entry.getValue());
+            // Return type of intermediate aggregation is the same as the input type.
+            RowExpression argumentExpr = aggregation.getCall().getArguments().get(0);
+            Type returnType = isExpression(argumentExpr) ? types.get(castToExpression(argumentExpr)) : argumentExpr.getType();
```

Comment:
it might be better to use the functionAndTypeRegistry to get the intermediate type similar to what we do for partial aggregations since that's the official source of truth for function types, but this is fine too - don't feel strongly. https://github.com/prestodb/presto/blob/69a924e701bec151e1ddaba5d55b7f7eb18df1ed/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java#L218-L236